jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -1122,6 +1122,56 @@ public static boolean hasScheme(String uri)
         return false;
     }
 
+    /**
+     * True if token is a <a href="https://www.rfc-editor.org/rfc/rfc3986">RFC3986</a> {@code reg-name}
+     *
+     * @param token the to test
+     * @return true if the token passes as a valid reg-name
+     */
+    public static boolean isRegName(String token)
+    {
+        /* reg-name ABNF is defined as :
+         *   reg-name      = *( unreserved / pct-encoded / sub-delims )
+         *   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+         *   pct-encoded   = "%" HEXDIG HEXDIG
+         *   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
+         *                   / "*" / "+" / "," / ";" / "="
+         */
+
+        if (token == null)
+            return true; // null token is considered valid
+
+        int length = token.length();
+        for (int i = 0; i < length; i++)
+        {
+            char c = token.charAt(i);
+            // unreserved
+            if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))
+                continue;
+            if (c == '-' || c == '.' || c == '_' || c == '~')
+                continue;
+            // pct-encoded
+            if (c == '%')
+            {
+                if (StringUtil.isHex(token, i + 1, 2))
+                {
+                    i += 2;
+                    continue;
+                }
+                else
+                    return false;
+            }
+            // sub-delims
+            if (c == '!' || c == '$' || c == '&' || c == '\'' || c == '(' || c == ')' ||
+                c == '*' || c == '+' || c == ',' || c == ';' || c == '=')
+            {
+                continue;
+            }
+            return false;
```

Comment:
Rather than 4 if, 3 of them with multiple comparisons.... it would be faster to have token lookup array.... in fact you could probably use `HttpTokens.TOKENS` and switch on the `Token`?    if not, then perhaps a secondary lookup in the same class?